rp2040-hal-examples/src/bin/spi_eh_bus.rs

```diff
@@ -0,0 +1,197 @@
+// Ensure we halt the program on panic (if we don't mention this crate it won't
+// be linked)
+use panic_halt as _;
+
+use rp2040_hal::gpio::PinState;
+use rp2040_hal::uart::{DataBits, StopBits, UartConfig};
+// Alias for our HAL crate
+use rp2040_hal as hal;
+
+// Some traits we need
+use core::fmt::Write;
+use hal::clocks::Clock;
+use hal::fugit::RateExtU32;
+
+// A shorter alias for the Peripheral Access Crate, which provides low-level
+// register access
+use hal::pac;
+
+/// The linker will place this boot block at the start of our program image. We
+/// need this to help the ROM bootloader get our code up and running.
+/// Note: This boot block is not necessary when using a rp-hal based BSP
+/// as the BSPs already perform this step.
+#[link_section = ".boot2"]
+#[used]
+pub static BOOT2: [u8; 256] = rp2040_boot2::BOOT_LOADER_GENERIC_03H;
+
+/// External high-speed crystal on the Raspberry Pi Pico board is 12 MHz. Adjust
+/// if your board has a different frequency
+const XTAL_FREQ_HZ: u32 = 12_000_000u32;
+
+/// Our Spi device driver
+/// We need to use a generic here (SPI), because we want our driver to work for any other
+/// microcontroller that implements the embedded-hal Spi traits
+struct MySpiDriver<SPI> {
+    spi: SPI,
+}
+
+/// When things go wrong, we could return Error(()) but that wouldn't help anyone troubleshoot
+/// so we'll create an error type with additional info to return.
+#[derive(Copy, Clone, Debug)]
+enum MyError<SPI> {
+    Spi(SPI),
+    // Add other errors for your driver here.
+}
+
+/// Implementation of the business logic for the remote Spi IC
+impl<SPI> MySpiDriver<SPI>
+where
+    SPI: SpiDevice,
+{
+    /// Construct a new instance of Spi device driver
+    pub fn new(spi: SPI) -> Self {
+        Self { spi }
+    }
+
+    /// Our hypothetical Spi device has a register at 0x20, that accepts a u8 value
+    pub fn set_value(&mut self, value: u8) -> Result<(), MyError<SPI::Error>> {
+        self.spi
+            .transaction(&mut [Operation::Write(&[0x20, value])])
+            .map_err(MyError::Spi)?;
+
+        Ok(())
+    }
+
+    /// Our hypothetical Spi device has a register at 0x90, that we can read a 2 byte value from
+    pub fn get_value(&mut self) -> Result<[u8; 2], MyError<SPI::Error>> {
+        let mut buf = [0; 2];
+        self.spi
+            .transaction(&mut [Operation::Write(&[0x90]), Operation::Read(&mut buf)])
+            .map_err(MyError::Spi)?;
+
+        Ok(buf)
+    }
+}
+
+/// Entry point to our bare-metal application.
+///
+/// The `#[rp2040_hal::entry]` macro ensures the Cortex-M start-up code calls this function
+/// as soon as all global variables and the spinlock are initialised.
+///
+/// The function configures the RP2040 peripherals, then performs some example
+/// SPI transactions, then goes to sleep.
+#[rp2040_hal::entry]
+fn main() -> ! {
+    // Grab our singleton objects
+    let mut pac = pac::Peripherals::take().unwrap();
```

Comment:
I've been trying to remove instances where we have a module and a variable of the same name (`pac`).

Reply:
do you have a preference here? the C programmer in me loves `p`, but maybe `peri` or `peripherals` would be nicer.

Reply:
The other option is to not import pac and use hal::pac.

Reply:
Renamed to p *and* use hal::pac